// react-textselect/dist/textselect.js
'use strict';

var React = require('react');
var lodashMap = require('lodash.map');

var TextSelect = React.createClass({
  displayName: 'TextSelect',

  propTypes: {
    options: React.PropTypes.any.isRequired,
    active: React.PropTypes.any.isRequired,
    onChange: React.PropTypes.func.isRequired,
    className: React.PropTypes.string
  },

  handleChange: function handleChange(event) {
    this.props.onChange(event, event.target.value, this.props.options[event.target.value]);
  },

  render: function render() {
    var _props = this.props;
    var options = _props.options;
    var active = _props.active;
    var className = _props.className;

    var classes = 'react-textselect';
    if (className) classes += ' ' + className;

    return React.createElement(
      'span',
      { className: classes },
      options[active],
      React.createElement(
        'select',
        { className: 'react-textselect-input', onChange: this.handleChange, value: active },
        lodashMap(options, function mapOptions(value, key) {
          return React.createElement(
            'option',
            { value: key, key: key },
            value
          );
        })
      )
    );
  }
});

module.exports = TextSelect;
